app/views/admin/inactive_users/_user_status.haml

```diff
@@ -0,0 +1,24 @@
+-# Show invitation status
+- if user.invitation_status == :pending_confirmation
+  Pending invitation confirmation.
+  %br
+- elsif user.invitation_status == :invitation_expired
+  Invitation expired.
+  %br
+-# Show reason for inactive status
+- if user.expired_at?
+  Account expired on #{user.expired_at}.
+- elsif user.expired?
+  Account expired due to inactivity. Last activity on #{user.last_activity_at}.
+- else
+  - version = user.versions.where(event: 'deactivate').order(created_at: :desc).first
+  - if version.nil?
+    Account deactivated.
+  - elsif can_audit_users? && version.whodunnit?
+    - deactivating_user = User.find_by_id(version.whodunnit.to_i)
+    - if deactivating_user.nil?
+      Account deactivated on #{version.created_at}.
+    - else
+      Account deactivated by #{deactivating_user.name} on #{version.created_at}.
+  - else
+    Account deactivated on #{version.created_at}.
```

Comment:
All this if-else nesting doesn't feel super clean. Is there a more rails-y way to do this? I am trying to avoid making the database queries until absolutely necessary.

Reply:
I'll think overnight on it, but I'd generally lean on putting a method on `User` and pass in the `current_user`.  Then you can do something like:
```
def deactivation_status(user),
  version = versions.where(event: 'deactivate').order(created_at: :desc)&.first
  return 'Account deactivated' unless version
  return 'Account deactivated on #{version.created_at}' unless user.can_audit_users? || version.whodunnit.blank?
 
  deactivating_user = User.find_by_id(version.whodunnit) # can you just call `version.user`?
  return "Account deactivated on #{version.created_at}" unless deactivating_user

  "Account deactivated by #{deactivating_user.name} on #{version.created_at}"
end
```

Reply:
and call it with `user.deactivation_status(current_user)`

Reply:
Hmm okay. It seems like it might cause confusion down the line to have this method as writing on the `User` because it will return `Account deactivated` even for users who are active or expired. I can make it return `Nil` or raise an exception if the user is not deactivated. Or, just have it return in the expiration information as a string as well. 

Just for my Rails learning... is another option to make this a helper method on the `users_controller` [like this](https://github.com/greenriver/hmis-warehouse/blob/d438ee1854d744fb3cede2ca2edba49710a57156/app/controllers/admin/edit_histories_controller.rb#L23-L29)? That makes slightly more sense to me than adding it to the `User` model, because in the controller we already have access to the current user, and it's the controllers responsibility to provide data to the view. I guess the downside of putting something in a controller is that other controllers/views won't be able to access that helper method.

Reply:
That works as well.  It looks like your goal with `name_of_whodunnit` is to just simplify the fetching of the user name of the whodunnit.  In that case, I think I'd put that method on `Version`.  You can do this with a monkey patch similar to this:
https://github.com/paper-trail-gem/paper_trail/issues/231#issuecomment-725493752  Then you could call `user.versions.last.name_of_whodunnit ` or something similar.